Three Dimension Graph Feature

ChrisMart21 opened the pull request:
# Description
This pull request introduces the 'Three Dimension Graph Feature' to the Open Energy Dashboard, aimed at enhancing visualization capabilities by allowing users to view and interact with data in three dimensions. The motivation behind this change is to provide an alternative modality for exploring energy consumption patterns over time.
Special thanks to @saulmendozaloera  @yyakhnis @Marty-Ronquillo for their contributions.

Fixes #895

## Type of change
- [ ] Note merging this changes the database configuration.
- [x] This change requires a documentation update

## Checklist
- [x] I have followed the [OED pull request](https://openenergydashboard.github.io/developer/pr.html) ideas
- [x] I have removed text in ( ) from the issue request
- [x] You acknowledge that every person contributing to this work has signed the [OED Contributing License Agreement](https://openenergydashboard.github.io/developer/cla.html) and each author is listed in the Description section.

ChrisMart21 committed:
Initial 3D Navigation Functionality

ChrisMart21 committed:
Merge remote-tracking branch 'upstream/development' into development

ChrisMart21 committed:
Merge branch 'OpenEnergyDashboard:development' into development

ChrisMart21 committed:
Merge remote-tracking branch 'upstream/development' into development

ChrisMart21 committed:
Merge remote-tracking
branch 'upstream/development' into development

ChrisMart21 committed:
ThreeDComponent required data format example

ChrisMart21 committed:
Initial Backend API Endpoint
  - Sample data retrieval on client end

ChrisMart21 committed:
Reading.js Fix

saulmendozaloera committed:
Merge pull request #1 from ChrisMart21/development

PR Test

yyakhnis committed:
Merge pull request #1 from ChrisMart21/development

Initial Back to front end api

ChrisMart21 committed:
Initial Redux ThreeD Graph Setup
    -types
    -state
    -actions
    -reducers

ChrisMart21 committed:
Redux State Integration for 3d Readings
	Selecting meters now pulls appropriate 3d data into redux state.

ChrisMart21 committed:
3D Graph initial logic
    - 3d Data now lives in redux state
    - 3dGraphs are now rendered from redux state

saulmendozaloera committed:
Begin implementing Mocha Test

saulmendozaloera committed:
Merge remote-tracking branch 'ChrisFork/development' into development for the purpose of being up to date with front end changes.

saulmendozaloera committed:
add threeDTests.js file

Commit:
Merge remote-tracking branch 'ChrisFork/development' into development

saulmendozaloera committed:
Changed naming of route path to be threeD rather than threed. Also added the algorithm in getThreeDReadings to detect missing data.

ChrisMart21 committed:
Initial Date Range Picker

saulmendozaloera committed:
Moment changes for XData and yData.

Commit:
Merge remote-tracking branch 'SaulRemote/development' into development

ChrisMart21 committed:
WIP

ChrisMart21 committed:
Merge remote-tracking branch 'SaulRemote/development' into threeDDev

ChrisMart21 committed:
Return Moment Instead

ChrisMart21 committed:
Merge remote-tracking branch 'upstream/development' into threeDDev

ChrisMart21 committed:
DateRangPicker Compatibility Functions

ChrisMart21 committed:
DateRange Implementation Update

ChrisMart21 committed:
Time Interval Bug Fix

ChrisMart21 committed:
ThreeD Component Updates

ChrisMart21 committed:
ThreeD Component Scaling and labeling updates.

ChrisMart21 committed:
Completed TODO for loading Spinner

ChrisMart21 committed:
Fix Single Line if parentheses

ChrisMart21 committed:
Update ThreeDReading for Moment.valueOf()

ChrisMart21 committed:
Fix ThreeDReading types
- corrected null scaling issues

ChrisMart21 committed:
Minor refactor to datePicker useSelector()
-Made use select more fine grained which seems to fix some fetch issues

ChrisMart21 committed:
ThreeDComponent Layout adjustments.

ChrisMart21 committed:
Initial Group Support & Help Text

ChrisMart21 committed:
3D Component Refactor.

ChrisMart21 committed:
Re-refactor of ThreeDComponent.

ChrisMart21 committed:
Initial Pills Concept

ChrisMart21 committed:
Pill Updates

ChrisMart21 committed:
Updates

ChrisMart21 committed:
3dState  Refactor

ChrisMart21 committed:
Readings updates

saulmendozaloera committed:
added an additional interger:sequenceNumber for frequency readings in the 3D graphics. Commit is done in order for front end to test.

ChrisMart21 committed:
Comments and Locale updates.

ChrisMart21 committed:
Merge remote branch 'SaulRemote/development'

ChrisMart21 committed:
Initial 3D reading support
-Dropdown for Precision added
-API endpoint updated to accept precision

ChrisMart21 committed:
Merge branch 'threeDSequence' into threeDPills

ChrisMart21 committed:
Merge branch 'threeDPills' into threeDDev

ChrisMart21 committed:
Merge remote-tracking branch 'upstream/development' into threeDDev

Marty-Ronquillo committed:
validate meter ID

Marty-Ronquillo committed:
Merge remote-tracking branch 'ChrisFork/threeDDev' into development

huss committed:
some changes but not yet optimized

huss committed:
okay version of 3D graphic DB code

This version supports multiple meters for groups. Uses special case for
every hour since much faster. Scales with the number of meters and
the number of points returned. It can take several seconds for 1 year
and every two hours. It probably needs more optimization but should
work reasonably and for first shot. Groups with lots of meters and/or
long times (1 year) may be slow.

ChrisMart21 committed:
Hole Algorithm refactor.

ChrisMart21 committed:
Merge remote-tracking branch 'SaulRemote/shl' into qMerge

ChrisMart21 committed:
Updates for updated sql parameters

ChrisMart21 committed:
Merge remote-tracking branch 'upstream/development' into qMerge

ChrisMart21 committed:
Readings Per Day

saulmendozaloera committed:
Adding @Huss's work/changes on: meter_3d_readings_unit and shrink_tsrange_to_meter_readings_by_day. Also added work on Reading.js and unitReading.js for group readings.

Marty-Ronquillo committed:
Validation for one meter on the 3D graph component
Working to make multiple meters work
Currently the code hangs when trying to
select another meter with the intent of having
multiple meters been passed through

ChrisMart21 committed:
Merge remote-tracking branch 'SaulRemote/development' into groups

ChrisMart21 committed:
Initial Groups

ChrisMart21 committed:
Merge branch 'groups' into qMerge

Marty-Ronquillo committed:
validation

yyakhnis committed:
8-5-2023 commit

ChrisMart21 committed:
Merge remote-tracking 'MartinFork/development'

ChrisMart21 committed:
Query Fix

yyakhnis committed:
8-5-2023 commit

yyakhnis committed:
commit 8/6/2023

ChrisMart21 committed:
Validation Fixes

ChrisMart21 committed:
ScrollBar CSS adjustment
-- Overflow allowed.

ChrisMart21 committed:
Clean console logs, add TODOS

ChrisMart21 committed:
yAxis tick/label overlap workaround.

ChrisMart21 committed:
Persistence Between Graphics

ChrisMart21 committed:
Help ToolTip alignment fix.

ChrisMart21 committed:
HoverText Translations

ChrisMart21 committed:
Double Fetch Issue fix.

ChrisMart21 committed:
Local State For Date Picker
-Simplify component and redux sate

ChrisMart21 committed:
3D ChartLink Support

ChrisMart21 committed:
Address 3D Additive Scaling Issues

saulmendozaloera committed:
Moved the validateThreeD functions outside of the router as well as renamed them based on the team's decision.

yyakhnis committed:
Route unit tests, Changes to Web unit tests, and corrections to bugs in Readings.js and unitReadings.js

ChrisMart21 committed:
Merge remote branch 'YashaRemote/development'
-- Route Tests
-- Web Test

yyakhnis committed:
Merge remote - Chris fork/testFeedback
- bug fixes/improvements to frontend

yyakhnis committed:
unit tests and debug
- added group 3D route test
- removed unnecessary console log
- fixed bug in getGroupThreeDReadings

ChrisMart21 committed:
Initial Area Normalization Support

ChrisMart21 committed:
Locale changes due to pr#958

huss committed:
Merge remote-tracking branch 'upstream/development' into dev

huss committed:
fix merge mistake

huss committed:
Merge remote-tracking branch 'origin/dev' into dev

ChrisMart21 committed:
Suppress 3D export data option

ChrisMart21 committed:
Merge branch 'dev' of https://github.com/ChrisMart21/OED-CSUMB-Capstone-Sum23 into dev

ChrisMart21 committed:
HoverText Corrections

ChrisMart21 committed:
Address code review issues
Updated
 - Formatting
 - Comments
 - JS Docs
 - Cleanup TODOS

ChrisMart21 committed:
Translations

saulmendozaloera committed:
Needed to merge work from Chris's fork in order to being needed edits per Huss request.

saulmendozaloera committed:
Moved comment lines for unitReadings.js and database.js. Also implemented logic that sends error 400 for 3D route when user picks dates greater than a year. lastly, removed duplicate validation fuctions.

saulmendozaloera committed:
Changed one of the query back to sequenceNumber.

ChrisMart21 committed:
Merge remote-tracking branch
'https://github.com/saulmendozaloera/OED/development' into dev

ChrisMart21 committed:
ReadingsPerDay -> ReadingInterval
- Renamed ReadingsPerDay for clarity.

ChrisMart21 committed:
Hole Algo Consolidation, and updates

ChrisMart21 committed:
Fix missing params

huss committed:
update comments & remove import

huss committed:
remove duplicate function declarations

huss committed:
add missed comment

huss committed:
remove blank line

saulmendozaloera committed:
Put comment back on ports that was used for debuggin. Else if statment edited to check for a duration that is greater than 1 year.

saulmendozaloera committed:
Removed readingsPerDay in comment lines for meterThreeDReadings and groupThreeDReadings.

huss committed:
add readings per day help

huss committed:
comment out debug port

huss committed:
update comment

huss committed:
avoid 3D points faster than meters provide

If the meter(s) have a reading frequency that is greater than
the requested time per point than it is now adjusted. No data returned
if greater than 12 hours.

ChrisMart21 committed:
Merge remote-tracking branch 'origin/shl' into dev

huss committed:
return special values if cannot do 3D

ChrisMart21 committed:
Merge remote-tracking branch 'origin/shl' into dev

ChrisMart21 committed:
Sample Low Freq Meter Approach

ChrisMart21 committed:
Low Freq Meter Changes/Fixes
- HoverText & label updates
- Translations
- Readings Per Day Updates

ChrisMart21 committed:
Fix duped Midpoint

huss committed:
update for 3D tests WIP

- Many 3D tests commented out for testing
- Fix validation of 3D to work if time unbounded
- Relocate reading test functions to util/
- Remove unused 3D test file
- Many new tests

huss committed:
Merge remote-tracking branch 'upstream/development' into shl

not building yet but merge conflicts gone

huss committed:
Package issue resolved/running

There were 7 packages missing and they were put back with npm installed.

huss committed:
3D uses new test data input

All 3D tests are working

huss committed:
more tests

< 1 day, flow, raw. fix freq. for 20 min group meters.

huss committed:
3D tests done/working

- added hole tests
- wording edits, format

huss committed:
careful read cleanup

ChrisMart21 committed:
Plotly TS.

ChrisMart21 committed:
JS Doc Fix & Naming

ChrisMart21 committed:
useSelector refactor for warnings

drkrupp committed:
Test Numero Dos

ChrisMart21 committed:
Merge remote branch 'upstream/development'

huss committed:
Merge remote-tracking branch 'huss/websiteUpdate' into pr/ChrisMart21/1063

huss committed:
#D tests updated to new way + other

- The new group input is used and standard kWh group included.
- Group insetion defaults displayable to false.
- The new file separation is used.
- L6 put into correct file.
- Misc. edits: removed unused imports, formatting, etc.
- The Design Doc is updated to include all these tests and changes.

huss commented:
I merged PR #1036 so the code could be consistent with those changes.  I also update the 3D tests and other test files as described in the commit. This PR is not expected to clear before the one merged so it should be fine. All changes should only impact test code. Any comments are welcome.

huss committed:
Merge remote-tracking branch 'origin/development' into pr/ChrisMart21/1063

drkrupp committed:
Added 3D rate limiting

drkrupp committed:
Merge remote-tracking branch 'DanielRemote/DanielBranch' into DanielBranch

drkrupp committed:
Removed Upstream Comment

ChrisMart21 committed:
Merge pull request #1 from drkrupp/DanielBranch

3D Rate Limiting

huss committed:
minor edits

- Update some comments
- precision -> readingInterval so consistently named
- Fix spelling of dateRangeCompatability
- remove condirional expressions where item already true/false
- Change variable name if single and not plural
- Formatting
- Increase ma

huss committed:
change 3D rate limit

Testing for 3D, esp. from rejection tests, was running too fast for the
rate limit so it was upped. Desire for a better solution is noted
in TODO.

huss commented:
Since I have also worked a little on this code and to move it along, I pushed minor edits, merged in development and then made some minor/modest changes. @ChrisMart21 can you look them over to be sure all is okay as I think I am done and all seems to work correctly.

huss committed:
Merge remote-tracking branch 'origin/development' into pr/ChrisMart21/1063

huss committed:
` -> ' in appropriate 3D test strings

huss committed:
remove duplicated state variable

huss committed:
translate pill titles

huss committed:
remove commented out code

huss committed:
remove unneeded require

huss committed:
update graph type tooltip for 3D

huss committed:
add link & text for date range help page

huss committed:
internationalize time of day label

huss committed:
Internationalize hover time

huss reviewed (commented):
Thanks to @ChrisMart21 for taking the lead on this significant and new feature along with  @saulmendozaloera @yyakhnis @Marty-Ronquillo who helped. This worked well and a code review found only minor items. I have made those edits and careful testing indicates that it all seems to work as desired. This is not being approved since I want to give the developers a chance to look over my changes. Unless issues are found, this is ready to merge from my point of view and I'll approve it.

huss commented:
This PR is ready to merge once the following happens:

- [x] PR #1036 clears. This PR includes those changes so it is preferred that the included PR clears first.
- [x] The recent merge conflict is resolved. The plan is to wait in case another merge causes any other issues.

I plan to do this as soon as the other needed items are done.

huss committed:
Merge remote-tracking branch 'origin/development' into pr/ChrisMart21/1063

huss commented:
As I'm about to merge this code, I want to especially thank @ChrisMart21 for their hard work, willingness to work longer than originally agreed and being the lead in getting this feature to be a reality.

huss reviewed (approved):
Everything is now ready and this works as expected.
